### DISCUSSION TITLE
Tower Defense - CI/CD

### DISCUSSION THREAD

**Discussion by Freaky2164** [Freaky2164/TowerDefense#49] [General]
Hi all,

for this weeks task of setting up a CI/CD pipeline for our project we unfortunately have bad news. It was not possible for us to set up a pipeline that builds our Unity game and runs our tests with any tool we tested. We want to give you a quick run down of what we tested and why it didnt work.

## Unity Build Automation
We looked into the CI/CD solution that Unity itself provides for their customers. The problem here is that this service would require us to pay a certain monthly fee.

## Jenkins
Our second approach was to set up a pipeline using Jenkins. This could potentially work because we could set everything up locally on a machine with Unity already installed. However the problem is that we do not have an additional server the pipeline could run on. Instead someone would have to manually trigger the pipeline to run on their machine. This means that we wouldn't have an automated pipeline that builds and runs our tests whenever someone changes something in our project. Thus defeating the whole purpose of the automation.

## Travis CI and CircleCI
We also looked into two other CI/CD solutions Travis CI and Circle CI. Both are free and it was possible for us connect them to our repository. Unfortunately they do not have any kind of support for Unity games, so we couldn't use them either.

## Game CI
The closest we got to a suitable solution was with help of Game CI. Game CI is an organization that tries to provide free CI/CD for game developers. It can be used with GitHub Actions, GitLab or CircleCI. For our project we looked into setting everything up with GitHub Actions and CircleCI. We follwed their documentation until we had to aquire a license from Unity that the pipeline needs in order to run correctly. For some reason, probably a bug somewhere, Unity doesn't want to give us this license file and so once again we couldn't claim victory after this long and frustrating battle.

<br>

But we don't want to end this blog on such a sad note. After thinking for a little while we came up with a solution that at least makes it possible to automatically test a part of our project. We plan to set up a pipeline that tests the C# scripts of our game which hold all of our game logic. That way we can make sure that the source code of our game is being tested and works as expected.

Greetings

Tower Defense Team

**Comment by Mscha1000** [comment]
Hi Tower Defense Team,

Thank you for sharing the update on your journey to set up a CI/CD pipeline for your Unity game. It's unfortunate to hear that you faced several challenges with the tools you explored.

Unity Build Automation seems promising, but the associated monthly fee didn't align with your project's requirements. Jenkins, although a potential solution, posed limitations due to the lack of an additional server to run the pipeline automatically. Travis CI and CircleCI, despite being free, didn't support Unity games directly, which proved to be a roadblock.

Your efforts with Game CI showed promise, but the struggle to acquire the necessary license from Unity hindered progress. It can be frustrating when unexpected bugs or issues arise, especially after investing time and energy.

However, it's great to see your team's resilience and ability to adapt. Focusing on automating the testing of C# scripts is a clever workaround. By ensuring the functionality of your game logic, you're taking a step towards maintaining code quality and reducing potential issues.

Though the journey may not have ended as expected, the creativity and problem-solving skills your team demonstrated are commendable. Keep up the great work, and I hope you find continued success in your development endeavors.

Best regards,
eicompany

**Comment by Mohmad-Naser-alnakeshbandi** [comment]
Hi Tower Defense Team,

I just wanted to drop a comment and express my appreciation for sharing the update on your journey to establish a CI/CD pipeline for your Unity game. It's unfortunate to hear about the challenges you encountered while exploring various tools. I have some talks with one of your team members about it.:)

I understand your concern regarding the Unity Build Automation's monthly fee not aligning with your project's requirements. Cost can be a significant factor, especially for indie developers or smaller teams. Additionally, the limitations of Jenkins without an additional server hindered its potential as a solution. It's also disappointing that Travis CI and CircleCI didn't directly support Unity games, creating a roadblock in your pipeline setup.

However, I'm impressed by your team's resilience and ability to adapt to the situation. Your decision to focus on automating the testing of C# scripts is a smart workaround. By ensuring the functionality of your game logic, you're taking a crucial step towards maintaining code quality and minimizing potential issues.

I understand that unexpected bugs and obstacles can be frustrating, especially after investing significant time and effort. But I want to emphasize that your creativity and problem-solving skills shine through in this situation. It's commendable how you've persevered and found alternative approaches to move forward.

Although the journey may not have unfolded as expected, I have no doubt that your team's dedication and adaptability will lead you to success. Keep up the great work, and I wish you continued progress in all your development endeavors.

Best regards,
Ghalia Team

**Comment by KevStr06** [comment]
Hi Tower Defense Team,

It's unfortunate to hear that you faced challenges while setting up a CI/CD pipeline for your Unity game. Building and testing a game can indeed be a complex task. However, it's great to see that you have found a partial solution by setting up a pipeline to test the C# scripts.

While it may not cover the entire game, focusing on testing the game logic through automated C# script tests is a step in the right direction. This way, you can ensure the core functionality of your game is tested automatically, reducing the risk of introducing bugs.

In the future, if you're still looking for a comprehensive CI/CD solution for building and testing your Unity game, you might consider exploring other options. The landscape of CI/CD tools is constantly evolving, and new solutions may emerge that better cater to game development needs.

Don't be disheartened by the challenges you faced. By continuously exploring and experimenting with different tools and approaches, you'll eventually find the right solution that works for your team. Keep up the great work, and best of luck with your ongoing development!

Best regards,
Team cyBuy
